# What is Numpy

NumPy is the fundamental package for scientific computing with Python. 
It is a package that provide high-performance vector, matrix and higher-dimensional data structures for Python. 
It is implemented in C and Fortran so when calculations are **vectorized**, performance is very good.

So, in a nutshell:

* a powerful Python extension for N-dimensional array
* a tool for integrating C/C++ and Fortran code
* designed for scientific computation: linear algebra and Signal Analysis

If you are a MATLAB&reg; user we recommend to read [Numpy for MATLAB Users](http://www.scipy.org/NumPy_for_Matlab_Users) and [Benefit of Open Source Python versus commercial packages](http://www.scipy.org/NumPyProConPage). 

I'm a supporter of the **Open Science Movement**, thus I humbly suggest you to take a look at the [Science Code Manifesto](http://sciencecodemanifesto.org/)

# Getting Started with Numpy Arrays

NumPy's main object is the **homogeneous** ***multidimensional array***. It is a table of elements (usually numbers), all of the same type. 

In Numpy dimensions are called **axes**. 

The number of axes is called **rank**. 

The most important attributes of an ndarray object are:

* **ndarray.ndim**     - the number of axes (dimensions) of the array. 
* **ndarray.shape**    - the dimensions of the array. For a matrix with n rows and m columns, shape will be (n,m). 
* **ndarray.size**     - the total number of elements of the array. 
* **ndarray.dtype**    - numpy.int32, numpy.int16, and numpy.float64 are some examples. 
* **ndarray.itemsize** - the size in bytes of elements of the array. For example, elements of type float64 has itemsize 8 (=64/8) 

To use `numpy` need to import the module it using of example:

In [1]:
import numpy as np  # naming import convention

### Terminology Assumption

In the `numpy` package the terminology used for vectors, matrices and higher-dimensional data sets is *array*. 

### Reference Documentation

* On the web: [http://docs.scipy.org](http://docs.scipy.org)/

* Interactive help:

In [8]:
np.array?

If you're looking for something

## Numpy Array Object

`NumPy` has a multidimensional array object called ndarray. It consists of two parts as follows:
   
   * The actual data
   * Some metadata describing the data
    
    
The majority of array operations leave the raw data untouched. The only aspect that changes is the metadata.

<img src="images/ndarray_with_details.png" />

## Creating `numpy` arrays

There are a number of ways to initialize new numpy arrays, for example from

* a Python list or tuples
* using functions that are dedicated to generating numpy arrays, such as `arange`, `linspace`, etc.

### From lists

For example, to create new vector and matrix arrays from Python lists we can use the `numpy.array` function.

In [15]:
v = np.array([1,2,3,4])

In [11]:
M = np.array([[1, 2], [3, 4]])
M

array([[1, 2],
       [3, 4]])

The `v` and `M` objects are both of the type `ndarray` that the `numpy` module provides.

In [12]:
print('Type of v: ', type(v))
print('Type of M: ', type(M))

Type of v:  <class 'numpy.ndarray'>
Type of M:  <class 'numpy.ndarray'>


The difference between the `v` and `M` arrays is only their shapes. 

To do so, we could use the `numpy.shape` function:

In [13]:
print('Shape of v: ', np.shape(v))
print('Shape of M: ', np.shape(M))

Shape of v:  (4,)
Shape of M:  (2, 2)


Alternatively, We can get information about the shape of an array by using the `ndarray.shape` **property** :

In [14]:
v.shape, M.shape

((4,), (2, 2))

Equivalently, we can get information about the **size** of the two `ndarrays`, namely the *total number of elements* in the array.

In [13]:
print('Size of v:', v.size)
print('Size of M:', M.size)

Size of v: 4
Size of M: 4


#### More properties of the `numpy array`

In [32]:
M.itemsize # bytes per element

8

In [33]:
M.nbytes # number of bytes

32

In [16]:
M.ndim # number of dimensions

2

## Using array-generating functions

For larger arrays it is inpractical to initialize the data manually, using explicit python lists. 

Instead we can use one of the many **functions** in `numpy` that generates arrays of different forms. 

Some of the more common are: 

* `np.arange`; 
* `np.linspace`; 
* `np.logspace`; 
* `np.mgrid`;
* `np.random.rand`;
* `np.diag`;
* `np.zeros`;
* `np.ones`;
* `np.empty`;
* `np.tile`.

### `np.arange`

In [17]:
x = np.arange(0, 10, 1) 
print(x)

[0 1 2 3 4 5 6 7 8 9]


In [18]:
# floating point step-wise range generatation
x = np.arange(-1, 1, 0.1)  
print(x)

[-1.00000000e+00 -9.00000000e-01 -8.00000000e-01 -7.00000000e-01
 -6.00000000e-01 -5.00000000e-01 -4.00000000e-01 -3.00000000e-01
 -2.00000000e-01 -1.00000000e-01 -2.22044605e-16  1.00000000e-01
  2.00000000e-01  3.00000000e-01  4.00000000e-01  5.00000000e-01
  6.00000000e-01  7.00000000e-01  8.00000000e-01  9.00000000e-01]


### `np.linspace` and `np.logspace`

In [18]:
# using linspace, both end points **ARE included**
np.linspace(0, 10, 25)

array([  0.        ,   0.41666667,   0.83333333,   1.25      ,
         1.66666667,   2.08333333,   2.5       ,   2.91666667,
         3.33333333,   3.75      ,   4.16666667,   4.58333333,
         5.        ,   5.41666667,   5.83333333,   6.25      ,
         6.66666667,   7.08333333,   7.5       ,   7.91666667,
         8.33333333,   8.75      ,   9.16666667,   9.58333333,  10.        ])

In [36]:
np.logspace(0, np.e**2, 10, base=np.e)

array([  1.00000000e+00,   2.27278564e+00,   5.16555456e+00,
         1.17401982e+01,   2.66829540e+01,   6.06446346e+01,
         1.37832255e+02,   3.13263169e+02,   7.11980032e+02,
         1.61817799e+03])

### `np.mgrid`

In [21]:
x, y = np.mgrid[0:5, 0:5]  # similar to meshgrid in MATLAB

In [22]:
x

array([[0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1],
       [2, 2, 2, 2, 2],
       [3, 3, 3, 3, 3],
       [4, 4, 4, 4, 4]])

In [23]:
y

array([[0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4]])

### `np.random.rand` & `np.random.randn`

In [1]:
import numpy as np
np.random.seed(4)

In [2]:
# uniform random numbers in [0,1]
R = np.random.rand(5,5)

In [3]:
R

array([[0.96702984, 0.54723225, 0.97268436, 0.71481599, 0.69772882],
       [0.2160895 , 0.97627445, 0.00623026, 0.25298236, 0.43479153],
       [0.77938292, 0.19768507, 0.86299324, 0.98340068, 0.16384224],
       [0.59733394, 0.0089861 , 0.38657128, 0.04416006, 0.95665297],
       [0.43614665, 0.94897731, 0.78630599, 0.8662893 , 0.17316542]])

In [3]:
R

array([[0.96702984, 0.54723225, 0.97268436, 0.71481599, 0.69772882],
       [0.2160895 , 0.97627445, 0.00623026, 0.25298236, 0.43479153],
       [0.77938292, 0.19768507, 0.86299324, 0.98340068, 0.16384224],
       [0.59733394, 0.0089861 , 0.38657128, 0.04416006, 0.95665297],
       [0.43614665, 0.94897731, 0.78630599, 0.8662893 , 0.17316542]])

In [25]:
# standard normal distributed random numbers
np.random.randn(5,5)

array([[ 0.65782724,  0.65168367,  0.58525852,  0.33781734, -0.00700978],
       [ 0.61574011,  0.59150639, -0.33797592, -0.2509655 ,  0.77237429],
       [-0.15693266, -0.38377945, -0.28140147,  0.90558314,  0.25437408],
       [-1.136108  ,  2.43964939,  0.28583627, -0.27540796, -0.57253111],
       [-0.79080395,  0.50525127,  2.1113386 , -0.33769711, -0.64914575]])

### `np.diag`

In [27]:
# a diagonal matrix
np.diag([1,2,3])

array([[1, 0, 0],
       [0, 2, 0],
       [0, 0, 3]])

In [5]:
# diagonal with offset from the main diagonal
np.diag([1,2,3], k=-1)[::-1]

array([[0, 0, 3, 0],
       [0, 2, 0, 0],
       [1, 0, 0, 0],
       [0, 0, 0, 0]])

### `np.eye`

In [7]:
# a diagonal matrix with ones on the main diagonal
np.eye(3, dtype='int')  # 3 is the 

array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 1]])

### `np.zeros` and `np.ones`

In [30]:
np.zeros((3,3))

array([[ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.]])

In [31]:
np.ones((3, 3))

array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.]])

### DIY

***Try by yourself*** the following commands:

    np.zeros((3,4))
    np.ones((3,4))
    np.empty((2,3))
    np.eye(5)
    np.diag(np.arange(5))
    np.tile(np.array([[6, 7], [8, 9]]), (2, 2))

## So, why is it useful then?

So far the `numpy.ndarray` looks awefully much like a Python **list** (or **nested list**). 

*Why not simply use Python lists for computations instead of creating a new array type?*

There are several reasons:

* Python lists are very general. 
    - They can contain any kind of object. 
    - They are dynamically typed. 
    - They do not support mathematical functions such as matrix and dot multiplications, etc. 
    - Implementing such functions for Python lists would not be very efficient because of the dynamic typing.
    
    
* Numpy arrays are **statically typed** and **homogeneous**. 
    - The type of the elements is determined when array is created.
    
    
* Numpy arrays are memory efficient.
    - Because of the static typing, fast implementation of mathematical functions such as multiplication and addition of `numpy` arrays can be implemented in a compiled language (C and Fortran is used).

In [1]:
import numpy as np

In [7]:
L = range(100000)

In [8]:
%timeit [i**2 for i in L]

31.9 ms ± 2.81 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [9]:
a = np.arange(100000)

In [10]:
%timeit a**2

58.2 µs ± 3.06 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [6]:
%timeit [element**2 for element in a]

228 µs ± 14.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


---

## Exercises

### Simple arrays

* Create simple one and two dimensional arrays. First, redo the examples
from above. And then create your own.

* Use the functions `len`, `shape` and `ndim` on some of those arrays and
observe their output.

### Creating arrays using functions

* Experiment with `arange`, `linspace`, `ones`, `zeros`, `eye` and `diag`.

* Create different kinds of arrays with random numbers.

* Try setting the seed before creating an array with random values 
    - *hint*: use `np.random.seed`

* Look at the function `np.empty`. What does it do? When might this be
useful?

---

## Basic Data Types

    bool             | This stores boolean (True or False) as a bit

    inti             | This is a platform integer (normally either int32 or int64)
    int8             | This is an integer ranging from -128 to 127
    int16            | This is an integer ranging from -32768 to 32767
    int32            | This is an integer ranging from -2 ** 31 to 2 ** 31 -1
    int64            | This is an integer ranging from -2 ** 63 to 2 ** 63 -1
    
    uint8            | This is an unsigned integer ranging from 0 to 255
    uint16           | This is an unsigned integer ranging from 0 to 65535
    uint32           | This is an unsigned integer ranging from 0 to 2 ** 32 - 1
    uint64           | This is an unsigned integer ranging from 0 to 2 ** 64 - 1

    float16          | This is a half precision float with sign bit, 5 bits exponent, and 10 bits mantissa
    float32          | This is a single precision float with sign bit, 8 bits exponent, and 23 bits mantissa
    float64 or float | This is a double precision float with sign bit, 11 bits exponent, and 52 bits mantissa
    complex64        | This is a complex number represented by two 32-bit floats (real and imaginary components)
    complex128       | This is a complex number represented by two 64-bit floats (real and imaginary components)
    (or complex)


## Numerical Types and Representation

The **numerical dtype** of an array should be selected very carefully, as it directly affects the numerical representation of elements, that is: 

   * the number of **bytes used; 
   * the *numerical range*

So, then: **What happens if I try to represent a number that is Out of range?**

Let's have a go with **integers**, i.e., `int8` and `uint8`

In [36]:
x = np.zeros(4, 'int8')  # Integer ranging from -128 to 127
x

array([0, 0, 0, 0], dtype=int8)

In [37]:
x[0] = 127
x

array([127,   0,   0,   0], dtype=int8)

In [38]:
x[0] = 128
x

array([-128,    0,    0,    0], dtype=int8)

In [39]:
x[1] = 129
x

array([-128, -127,    0,    0], dtype=int8)

In [40]:
x[2] = 257  # i.e. (128 x 2) + 1
x

array([-128, -127,    1,    0], dtype=int8)

In [41]:
ux = np.zeros(4, 'uint8')  # Integer ranging from 0 to 255
ux

array([0, 0, 0, 0], dtype=uint8)

In [42]:
ux[0] = 255
ux[1] = 256
ux[2] = 257
ux[3] = 513  # (256 x 2) + 1
ux

array([255,   0,   1,   1], dtype=uint8)

---

# Exercises - Basic Numpy

## Ex 1. Create an array containing integers from $2$ to $2^6$

#### Hint: use Python `range` function

## 1.1 Print `ndarray` attributes and properties
(e.g. `type`, `dtype`, `shape...`) using previous on

## Ex 2. Create a 3x3 Matrix array and fill it with integer numbers

## Ex3: Create a Matrix of any size and fill it with random numbers

### HInt: take a look at `numpy.random.rand`

---

# Indexing

### Setting up the data

In [56]:
np.random.seed(42)  # Setting the random seed

In [57]:
# a vector: the argument to the array function is a Python list
v = np.random.rand(10)
v

array([0.37454012, 0.95071431, 0.73199394, 0.59865848, 0.15601864,
       0.15599452, 0.05808361, 0.86617615, 0.60111501, 0.70807258])

In [58]:
# a matrix: the argument to the array function is a nested Python list
M = np.random.rand(10, 2)
M

array([[0.02058449, 0.96990985],
       [0.83244264, 0.21233911],
       [0.18182497, 0.18340451],
       [0.30424224, 0.52475643],
       [0.43194502, 0.29122914],
       [0.61185289, 0.13949386],
       [0.29214465, 0.36636184],
       [0.45606998, 0.78517596],
       [0.19967378, 0.51423444],
       [0.59241457, 0.04645041]])

We can index elements in an array using the square bracket and indices:

In [59]:
# v is a vector, and has only one dimension, taking one index
v[0]

0.3745401188473625

In [60]:
# M is a matrix, or a 2 dimensional array, taking two indices 
M[1,1]

0.21233911067827616

If we omit an index of a multidimensional array it returns the whole row (or, in general, a N-1 dimensional array) 

In [61]:
M[1] 

array([0.83244264, 0.21233911])

The same thing can be achieved with using `:` instead of an index: 

In [62]:
M[1,:] # row 1

array([0.83244264, 0.21233911])

In [63]:
M[:,1] # column 1

array([0.96990985, 0.21233911, 0.18340451, 0.52475643, 0.29122914,
       0.13949386, 0.36636184, 0.78517596, 0.51423444, 0.04645041])

We can assign new values to elements in an array using indexing:

In [64]:
M[0,0] = 1

In [65]:
M

array([[1.        , 0.96990985],
       [0.83244264, 0.21233911],
       [0.18182497, 0.18340451],
       [0.30424224, 0.52475643],
       [0.43194502, 0.29122914],
       [0.61185289, 0.13949386],
       [0.29214465, 0.36636184],
       [0.45606998, 0.78517596],
       [0.19967378, 0.51423444],
       [0.59241457, 0.04645041]])

In [66]:
# also works for rows and columns
M[1,:] = 0
M[:,1] = -1

In [67]:
M

array([[ 1.        , -1.        ],
       [ 0.        , -1.        ],
       [ 0.18182497, -1.        ],
       [ 0.30424224, -1.        ],
       [ 0.43194502, -1.        ],
       [ 0.61185289, -1.        ],
       [ 0.29214465, -1.        ],
       [ 0.45606998, -1.        ],
       [ 0.19967378, -1.        ],
       [ 0.59241457, -1.        ]])

## Index slicing

Index slicing is the technical name for the syntax `M[lower:upper:step]` to extract part of an array:

In [68]:
a = np.array([1,2,3,4,5])
a

array([1, 2, 3, 4, 5])

In [69]:
a[1:3]

array([2, 3])

Array slices are **mutable**: if they are assigned a new value the original array from which the slice was extracted is modified:

In [70]:
a[1:3] = [-2,-3]

a

array([ 1, -2, -3,  4,  5])

* We can omit any of the three parameters in `M[lower:upper:step]`:

In [71]:
a[::] # lower, upper, step all take the default values

array([ 1, -2, -3,  4,  5])

In [72]:
a[::2] # step is 2, lower and upper defaults to the beginning and end of the array

array([ 1, -3,  5])

In [73]:
a[:3] # first three elements

array([ 1, -2, -3])

In [74]:
a[3:] # elements from index 3

array([4, 5])

* Negative indices counts from the end of the array (positive index from the begining):

In [75]:
a = np.array([1,2,3,4,5])

In [76]:
a[-1] # the last element in the array

5

In [77]:
a[-3:] # the last three elements

array([3, 4, 5])

* Index slicing works exactly the same way for multidimensional arrays:

In [78]:
A = np.array([[n+m*10 for n in range(5)] 
              for m in range(5)])
A

array([[ 0,  1,  2,  3,  4],
       [10, 11, 12, 13, 14],
       [20, 21, 22, 23, 24],
       [30, 31, 32, 33, 34],
       [40, 41, 42, 43, 44]])

In [79]:
# a block from the original array
A[1:4, 1:4]

array([[11, 12, 13],
       [21, 22, 23],
       [31, 32, 33]])

In [80]:
# strides
A[::2, ::2]

array([[ 0,  2,  4],
       [20, 22, 24],
       [40, 42, 44]])

## Fancy indexing

Fancy indexing is the name for when an array or list is used in-place of an index: 

In [81]:
row_indices = [1, 2, 3]
A[row_indices]

array([[10, 11, 12, 13, 14],
       [20, 21, 22, 23, 24],
       [30, 31, 32, 33, 34]])

In [82]:
col_indices = [1, 2, -1] # remember, index -1 means the last element
A[row_indices, col_indices]

array([11, 22, 34])

* We can also index **masks**: 

    - If the index mask is an Numpy array of with data type `bool`, then an element is selected (True) or not (False) depending on the value of the index mask at the position each element: 

In [83]:
b = np.array([n for n in range(5)])
b

array([0, 1, 2, 3, 4])

In [84]:
row_mask = np.array([True, False, True, False, False])
b[row_mask]

array([0, 2])

* Alternatively:

In [85]:
# same thing
row_mask = np.array([1,0,1,0,0], dtype=bool)
b[row_mask]

array([0, 2])

This feature is very useful to conditionally select elements from an array, using for example comparison operators:

In [11]:
x = np.arange(0, 10, 0.5)
x

array([0. , 0.5, 1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. , 5.5, 6. ,
       6.5, 7. , 7.5, 8. , 8.5, 9. , 9.5])

In [12]:
mask = (5 < x)

mask

array([False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [13]:
x[mask]

array([5.5, 6. , 6.5, 7. , 7.5, 8. , 8.5, 9. , 9.5])

Alternatively, we can use the condition (mask) array directly within brackets to index the array

In [14]:
x[(5 < x)]

array([5.5, 6. , 6.5, 7. , 7.5, 8. , 8.5, 9. , 9.5])

---

# Exercises on Indexing

Index slicing is the technical name for the syntax `M[lower:upper:step]` to extract part of an array

## Ex 3.1 

Generate a three-dimensional array of any size containing random numbers taken from an uniform distribution (_guess the numpy function in `np.random`_). Then print out separately the first entry along the three axis (i.e. `x, y, z`)  


#### Hint: Slicing with numpy arrays works quite like Python lists

## Ex 3.2

Create a vector and print out elements in reverse order

#### Hint: Use slicing for this exercise

## Ex 3.3

Generate a $7 \times 7$ matrix and replace all the elements in odd rows and even columns with `1`.

#### Hint: Use slicing to solve this exercise!

#### Note: Take a look at the original matrix, then.

**Use fancy indexing** to get all the elements of the previous matrix that are equals to `1`

## Ex 3.4 

Generate a `10 x 10` matrix of numbers `A`. Then, generate a numpy array of integers in range `1-9`. Pick `5` random values (with no repetition) from this array and use these values to extract rows from the original matrix `A`.

## Ex 3.5 

Repeat the previous exercise but this time extract columns from `A`

## Ex 3.6

Generate an array of numbers from `0` to `20` with step `0.5`. 
Extract all the values greater than a randomly generated number in the same range.

#### Hint: Try to write the condition as an expression and save it to a variable. Then, use this variable in square brackets to index.... this is when the magic happens!